language/vm/vm-runtime/src/txn_executor.rs

```diff
@@ -162,16 +147,99 @@ where
         Ok(())
     }
 
+    /// Execute a function.
+    /// `module` is an identifier for the name the module is stored in. `function_name` is the name
+    /// of the function. If such function is found, the VM will execute this function with arguments
+    /// `args`. The return value will be placed on the top of the value stack and abort if an error
+    /// occurs.
+    pub fn execute_function(
+        &mut self,
+        runtime: &VMRuntime,
+        state_view: &dyn StateView,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        args: Vec<Value>,
+    ) -> VMResult<()> {
+        runtime.execute_function(
+            state_view,
+            &mut self.interpreter_context,
+            &self.txn_data,
+            &self.gas_schedule,
+            module,
+            function_name,
+            args,
+        )
+    }
+
+    /// Entrypoint into the interpreter. All external calls need to be routed through this
+    /// function.
+    pub(crate) fn execute_script(
+        &mut self,
+        runtime: &VMRuntime,
+        state_view: &dyn StateView,
+        script: Vec<u8>,
+        args: Vec<TransactionArgument>,
+    ) -> VMResult<()> {
+        runtime.execute_script(
+            state_view,
+            &mut self.interpreter_context,
+            &self.txn_data,
+            &self.gas_schedule,
+            script,
+            convert_txn_args(args),
+        )
+    }
+
+    /// Verifies that a `CompiledModule` is valid and returns the `ModuleId` if successful.
+    pub(crate) fn publish_module(
+        &mut self,
+        module: &[u8],
+        runtime: &VMRuntime,
+        state_view: &dyn StateView,
+    ) -> VMResult<ModuleId> {
+        runtime.publish_module(
+            module,
+            state_view,
+            &mut self.interpreter_context,
+            &self.txn_data,
+            self.gas_schedule,
+        )
+    }
+
+    /// Execute a function with the sender set to `sender`, restoring the original sender afterward.
+    /// This should only be used in the logic for generating the genesis block.
+    #[allow(non_snake_case)]
+    pub fn execute_function_with_sender_FOR_GENESIS_ONLY(
+        &mut self,
+        runtime: &VMRuntime,
+        state_view: &dyn StateView,
+        address: AccountAddress,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        args: Vec<Value>,
+    ) -> VMResult<()> {
+        let old_sender = self.txn_data.sender;
+        self.txn_data.sender = address;
+        let res = self.execute_function(runtime, state_view, module, function_name, args);
```

Comment:
@runtian-zhou had a good suggestion in the past that I forgot to implement: we should typecheck the args using something like the ` if !verify_actuals(main.signature(), &args)` code in `execute_script` to avoid mistakes in generating genesis. Would you mind adding a TODO for this?

Reply:
`execute_function` verifies the arguments not sure if we want the TODO, do we?